mono/metadata/class-internals.h

```diff
@@ -1006,33 +1006,29 @@ MonoClass* mono_class_try_get_##shortname##_class (void);
 
 #define GENERATE_GET_CLASS_WITH_CACHE(shortname,name_space,name) \
 MonoClass*	\
-mono_class_get_##shortname##_class (void)	\
-{	\
-	static MonoClass *tmp_class;	\
-	MonoClass *klass = tmp_class;	\
-	if (!klass) {	\
+mono_class_get_##shortname##_class (void)							\
+{												\
+	MONO_STATIC_POINTER_INIT (MonoClass, klass)						\
 		klass = mono_class_load_from_name (mono_defaults.corlib, name_space, name);	\
-		mono_memory_barrier ();	\
-		tmp_class = klass;	\
-	}	\
-	return klass;	\
+	MONO_STATIC_POINTER_INIT_END (MonoClass, klass)						\
+	return klass;										\
 }
 
-#define GENERATE_TRY_GET_CLASS_WITH_CACHE(shortname,name_space,name) \
-MonoClass*	\
-mono_class_try_get_##shortname##_class (void)	\
-{	\
-	static volatile MonoClass *tmp_class;	\
-	static volatile gboolean inited;	\
-	MonoClass *klass = (MonoClass *)tmp_class;	\
-	mono_memory_barrier ();	\
-	if (!inited) {	\
+#define GENERATE_TRY_GET_CLASS_WITH_CACHE(shortname,name_space,name)				\
+MonoClass*											\
+mono_class_try_get_##shortname##_class (void)							\
```

Comment:
This is subtle but incorrect.
This macro previously would try only approximately once and cache failure.
This new form repeatedly retries instead.

Reply:
The new version is wrong.  The `mono_class_try_get_xxx_class` functions should cache NULL as "initialized, but the class wasn't found"  (this can be used if a class is linked away by the managed linker, for example).  The original version would only try to load the class "once" (modulo concurrency) - your version will hit the slow path every time if the class isn't found.